storage/norcow_blockwise.h

```diff
@@ -0,0 +1,308 @@
+  *pos = offset;
+
+  const void *l = norcow_ptr(sector, *pos, NORCOW_LEN_LEN);
+  if (l == NULL) return secfalse;
+  memcpy(len, l, sizeof(uint16_t));
+
+  *pos += NORCOW_LEN_LEN;
+  const void *k = norcow_ptr(sector, *pos, NORCOW_KEY_LEN);
+  if (k == NULL) {
+    return secfalse;
+  }
+
+  if (*len <= NORCOW_SMALL_ITEM_SIZE) {
+    memcpy(key, k, sizeof(uint16_t));
+    if (*key == NORCOW_KEY_FREE) {
+      return secfalse;
+    }
+    *pos += NORCOW_KEY_LEN;
+  } else {
+    *pos += (NORCOW_KEY_LEN + NORCOW_SMALL_ITEM_SIZE);
+
+    const void *flg = norcow_ptr(sector, *pos, NORCOW_VALID_FLAG_LEN);
+    if (flg == NULL) {
+      return secfalse;
+    }
+
+    *pos += NORCOW_VALID_FLAG_LEN;
+    if (*((const uint8_t *)flg) == 0) {
+      // Deleted item.
+      *key = NORCOW_KEY_DELETED;
+    } else {
+      memcpy(key, k, sizeof(uint16_t));
+      if (*key == NORCOW_KEY_FREE) {
+        return secfalse;
+      }
+    }
+  }
+
+  *val = norcow_ptr(sector, *pos, *len);
+  if (*val == NULL) return secfalse;
+  *pos = FLASH_ALIGN(*pos + *len);
+  return sectrue;
+}
+
+void norcow_delete_head(const flash_area_t *area, uint32_t len,
+                        uint32_t *val_offset) {
+  ensure(flash_unlock_write(), NULL);
+  // Move to the beginning of the block.
+  if (len <= NORCOW_SMALL_ITEM_SIZE) {
+    // Will delete the entire small item, setting the length to 0
+    *val_offset -= NORCOW_LEN_LEN + NORCOW_KEY_LEN;
+  } else {
+    // Will update the flag to indicate that the old item has been deleted.
+    // Deletes a portion of old item data too.
+    *val_offset -= NORCOW_VALID_FLAG_LEN;
+  }
+
+  flash_block_t block = {0};
+  ensure(flash_area_write_block(area, *val_offset, block), NULL);
+
+  // Move to the next block.
+  *val_offset += FLASH_BLOCK_SIZE;
+  ensure(flash_lock_write(), NULL);
+}
+
+static secbool flash_area_write_bytes(const flash_area_t *area, uint32_t offset,
+                                      uint16_t dest_len, const void *val,
+                                      uint16_t len) {
+  if (val == NULL || dest_len != len) {
+    return secfalse;
+  }
+
+  uint8_t *data = (uint8_t *)val;
+
+  while (len > 0) {
+    flash_block_t block = {0};
+
+    uint16_t offset_in_block = offset % FLASH_BLOCK_SIZE;
+
+    memcpy(block,
+           flash_area_get_address(area, offset - offset_in_block,
+                                  FLASH_BLOCK_SIZE),
+           FLASH_BLOCK_SIZE);
+
+    int32_t len_to_write = len > (FLASH_BLOCK_SIZE - offset_in_block)
+                               ? (FLASH_BLOCK_SIZE - offset_in_block)
+                               : len;
+
+    memcpy(&((uint8_t *)block)[offset_in_block], data, len_to_write);
+
+    if (sectrue !=
+        flash_area_write_block(area, offset - offset_in_block, block)) {
+      return secfalse;
+    }
+
+    offset += len_to_write;
+    data += len_to_write;
+    len -= len_to_write;
+  }
+
+  return sectrue;
+}
```

Comment:
I think it doesn't make sense to implement this function for the blockwise flash access, because the only use-case that I am aware of is updating the PIN counter in storage, which wouldn't work with this anyway. In fact, we should get rid of the hardened PIN counter implementation at least for the U5 and maybe even for the older models.

Reply:
One example when this actually applies is when changing pin: the `PIN_NOT_SET_KEY` item gets overwritten with same value, so if this is not implemented the value gets deleted and re-added in norcow. Maybe its not worth implementing anyways, also this specific case can be probably handled on `storage.c` level. On the other hand, for similar future situations this might be good to keep. I will let you decide, i think its fine either way and if go without this, readding it in future, if the need arises, should not be a problem i believe.

Reply:
> the PIN_NOT_SET_KEY item gets overwritten with same value,
> On the other hand, for similar future situations this might be good to keep.

I think that those situations can be covered by something like this:
```c
return memcmp(val, flash_area_get_address(area, offset, dest_len), len) == 0 ? sectrue : secfalse;
```
Any additional complexity seems unwarranted.

Reply:
using simplified version in https://github.com/trezor/trezor-firmware/commit/f625f87fb56badc70fb6a47ff6ab4574aad0aed6